### PR TITLE
File permission transfer

### PR DESCRIPTION
**Pull request by dliang36**
Implementation and test for keeping file permission during transfer v0.1

Design decision:
1. Get permission from the same st_mode that is used to get file size, and store permission in each metadata entry of DirectorySourceQueue, in order to avoid calling stat another time.
2. Default permission is 512 (01000). When perm in WdtFileInfo is default, library behavior is not changed. 

Test:
1. Encode and decode in Protocol
2. Print src and dst file permissions to file and compare in E2E

TODO:
1. Shall we consider dir permission?
2. What if blockDetails.allocationStatus == TO_BE_DELETED but receiver's file permission is owner only?
3. What if sender side allows only owner's read? 
4. permission changed – only chmod the file. Don’t trans again

### CODE REVIEW

File: ByteSource.h [hunk 1]
```diff
@@ -44,6 +44,8 @@ struct SourceMetaData {
   FileAllocationStatus allocationStatus{NOT_EXISTS};
   /// if there is a size mismatch, this is the previous sequence id
   int64_t prevSeqId{0};
+  /// file permission.
+  int64_t permission{512};
```
Comment (uddipta) [inline ByteSource.h]: This can be int32_t.

File: Protocol.cpp [hunk 2]
```diff
@@ -33,6 +33,7 @@ const int Protocol::INCREMENTAL_TAG_VERIFICATION_VERSION = 25;
 const int Protocol::DELETE_CMD_VERSION = 26;
 const int Protocol::VARINT_CHANGE = 27;
 const int Protocol::HEART_BEAT_VERSION = 29;
+const int Protocol::KEEP_PERMISSION = 30;
```
Comment (uddipta) [inline Protocol.cpp]: PRESERVE_PERMISSION

File: Protocol.h [hunk 3]
```diff
@@ -218,6 +218,8 @@ struct BlockDetails {
   FileAllocationStatus allocationStatus{NOT_EXISTS};
   /// seq-id of previous transfer, only valid if there is a size mismatch
   int64_t prevSeqId{0};
+  /// file permission in decimal.
+  int64_t permission{512};
```
Comment (uddipta) [inline Protocol.h]: int32_t

File: CMakeLists.txt [hunk 4]
```diff
@@ -22,7 +22,7 @@ cmake_minimum_required(VERSION 3.2)
 # There is no C per se in WDT but if you use CXX only here many checks fail
 # Version is Major.Minor.YYMMDDX for up to 10 releases per day (X from 0 to 9)
 # Minor currently is also the protocol version - has to match with Protocol.cpp
-project("WDT" LANGUAGES C CXX VERSION 1.29.1703251)
+project("WDT" LANGUAGES C CXX VERSION 1.30.1703270)
```
Comment (uddipta) [inline CMakeLists.txt]: You will probably have to change the version to 31, as I am using 30 for an encryption fix.

File: WdtTransferRequest.h [hunk 5]
```diff
@@ -35,6 +35,12 @@ struct WdtFileInfo {
   /// Whether read should be done using o_direct. If fd is set, this flag will
   /// be set automatically to match the fd open mode
   bool directReads{false};
+  /// File permission in decimal. When smaller than oct777 (dec512), the
+  /// permission will be passed during transfer.
+  int64_t permission{512};
```
Comment (uddipta) [inline WdtTransferRequest.h]: Why this check against 512? Can't the default value be 0 and wdt always populates it to the right value in DirectorySourceQueue?

File: util/DirectorySourceQueue.cpp [hunk 6]
```diff
@@ -360,7 +366,9 @@ bool DirectorySourceQueue::explore() {
               !std::regex_match(newRelativePath, includeRegex)) {
             continue;
           }
-          WdtFileInfo fileInfo(newRelativePath, fileStat.st_size, directReads_);
+          int perm = fileStat.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO);
```
Comment (uddipta) [inline util/DirectorySourceQueue.cpp]: We need this logic also in enqueueFiles() method. If the users provide a list of files, we use that method to populate the queue.

File: util/FileCreator.cpp [hunk 7]
```diff
@@ -197,7 +198,16 @@ int FileCreator::openExistingFile(ThreadCtx &threadCtx,
   return res;
 }
 
-int FileCreator::createFile(ThreadCtx &threadCtx, const string &relPathStr) {
+int openWithPerm(const char *path, int oFlags, int perm) {
+  if (perm >= 512) {
```
Comment (uddipta) [inline util/FileCreator.cpp]: I am not sure we need this check against 512. Just set the default value of BlockDetails::permission to 0644.

**Comment by uddipta** [comment]
Regarding your questions, all of them are valid. But, let's try to resolve those one by one in separate diffs.

File: util/DirectorySourceQueue.cpp (hunk 6 above)
Comment (dliang36) [inline util/DirectorySourceQueue.cpp]: Oh my previous understanding was: we allow the user to specify a permission, which will be used during createFile. If no valid permission is specified in the request, we will create a 644. (That's why my default permission value is 512. Cuz 0 is still a theoretically valid permission) But I think you are right. It is better to always read the permission from the given file.

Comment (uddipta) [inline util/DirectorySourceQueue.cpp]: nit: const int perm

File: util/DirectorySourceQueue.cpp [hunk 8]
```diff
@@ -557,14 +566,15 @@ bool DirectorySourceQueue::enqueueFiles() {
...
+    }
     if (info.fileSize < 0) {
-      struct stat fileStat;
-      if (stat(fullPath.c_str(), &fileStat) != 0) {
-        WPLOG(ERROR) << "stat failed on path " << fullPath;
-        return false;
-      }
       info.fileSize = fileStat.st_size;
     }
+    info.permission = fileStat.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO);
```
Comment (uddipta) [inline util/DirectorySourceQueue.cpp]: Can we have a helper function for getting the permission from the mode? Don't like that we have two places with this constants.

int getPermission(int mode) {
  return mode & (S_IRWXU | S_IRWXG | S_IRWXO);
}

File: util/FileCreator.h [hunk 9]
```diff
@@ -108,7 +108,8 @@ class FileCreator {
    *
    * @return          file descriptor or -1 on error
    */
-  int createFile(ThreadCtx &threadCtx, const std::string &relPath);
+  int createFile(ThreadCtx &threadCtx,
+                 const std::string &relPath, int64_t permission);
```
Comment (uddipta) [inline util/FileCreator.h]: int32_t permission

File: test/wdt_e2e_simple_test.sh [hunk 10]
```diff
@@ -137,7 +137,13 @@ if [ $DO_VERIFY -eq 1 ] ; then
     echo "Should be no diff"
     (cd $DIR; diff -u src.md5s dst.md5s)
     STATUS=$?
+    if [ $STATUS -eq 0 ] ; then
+      (cd $DIR/src ; ( find . -type f -printf %f%m\\n | sort ) > ../src.perm )
```
Comment (jjleng) [inline test/wdt_e2e_simple_test.sh]: `find . -type f -printf %f%m\\n` doesn't work on Mac.  Can you also test files with permissions other than 644?

File: Protocol.h [hunk 11]
```diff
@@ -300,10 +304,10 @@ class Protocol {
...
   /// Max size of sender or receiver id
   static constexpr int64_t kMaxTransferIdLength = 50;
-  /// 1 byte for cmd, 2 bytes for file-name length, Max size of filename, 4
-  /// variants(seq-id, data-size, offset, file-size), 1 byte for flag, 10 bytes
-  /// prev seq-id
-  static constexpr int64_t kMaxHeader = 1 + 2 + PATH_MAX + 4 * 10 + 1 + 10;
+  /// 1 byte for cmd, 2 bytes for file-name length, Max size of filename, 5
+  /// variants(seq-id, data-size, offset, file-size, permission), 1 byte for
+  /// flag, 10 bytes prev seq-id
+  static constexpr int64_t kMaxHeader = 1 + 2 + PATH_MAX + 5 * 10 + 2 + 1 + 10;
```
Comment (jjleng) [inline Protocol.h]: No comments for the extra `+ 2`. Where does the 2 extra bytes come from?

File: test/wdt_e2e_simple_test.sh (hunk 10 above)
Comment (dliang36) [inline test/wdt_e2e_simple_test.sh]: Currently within the /tmp/wdtTest_[usrname], there are several files with 664, which include ".a" and ".b" if I remember correctly. Regarding Mac... I am not sure whether we need to consider that. @uddipta Could you clarify this? Similarly I think stat() might not work on Windows.

File: Protocol.h (hunk 11 above)
Comment (dliang36) [inline Protocol.h]: My neglect. I used that for permission before this commit. Should have been deleted. Thank you for pointing out.

File: test/wdt_e2e_simple_test.sh (hunk 10 above)
Comment (uddipta) [inline test/wdt_e2e_simple_test.sh]: 1) we do support Mac. We run open source continuous builds here https://travis-ci.org/facebook/wdt
2) In e2e simple test, I don't see any files having permissions other than 644.

So, we need to add a file with different permission for testing.

As Jason mentioned before, current find cmd does not work on mac. So, we can disable the test for now on Mac. You can do that using an environment variable. We can have an environment variable like SKIP_PERMISSION_TEST and set it in https://github.com/facebook/wdt/blob/master/build/travis_osx.sh.

If you have access to a mac, you can also try to make the test work for mac. Right now, find is failing for mac, but we are not checking the return code. Both, src.perm and dst.perm are empty, so the test is passing.

Also, run clangformat.sh once.

Every time you update this PR, you should see a travis build/test triggered.

Comment (dliang36) [inline test/wdt_e2e_simple_test.sh]: 1. OK. I will add a environment var for that. Sorry but I do not have a mac at hand...

2. Well... there are...
If I just change this line https://github.com/facebook/wdt/pull/155/files#diff-a77a3101b61bf7da9963a0791dba5bb6R247 to return 0644, 
The following tests FAILED:
	 10 - FileReaderTests (Failed)
	 12 - WdtBasicE2E (Failed)
	 19 - WdtSimpleOdirectTest (Failed)
For 12 & 19, we have things like this:
Should be no diff
--- src.perm	2017-04-05 14:56:37.893425997 -0700
+++ dst.perm	2017-04-05 14:56:37.893425997 -0700
@@ -1,7 +1,7 @@
-a664
-c664
-file1664
-file1664
+a644
+c644
+file1644
+file1644

(Part of "ls -l" output on the src dir is here.)
screenshot from 2017-04-05 15-05-34

Comment (uddipta) [inline test/wdt_e2e_simple_test.sh]: It is better to chmod a file to have permission like 777. Just do that for file1.

Regarding adding test for mac, we will add it later.

File: ReceiverThread.cpp [hunk 12]
```diff
@@ -6,13 +6,13 @@
...
  * of patent rights can be found in the PATENTS file in the same directory.
  */
-#include <wdt/ReceiverThread.h>
 #include <folly/Bits.h>
 #include <folly/Checksum.h>
 #include <folly/Conv.h>
 #include <folly/Memory.h>
 #include <folly/ScopeGuard.h>
 #include <folly/String.h>
+#include <wdt/ReceiverThread.h>
```
Comment (uddipta) [inline ReceiverThread.cpp]: Revert this one too.

"The associated header file of .cpp files should be included before any other includes. (This helps catch missing header file dependencies in the .h)"

This is the lint error I'm seeing.

File: Reporting.cpp [hunk 13]
```diff
@@ -6,9 +6,9 @@
  * LICENSE file in the root directory of this source tree. An additional grant
  * of patent rights can be found in the PATENTS file in the same directory.
  */
-#include <wdt/Reporting.h>
 #include <folly/String.h>
 #include <wdt/Protocol.h>
+#include <wdt/Reporting.h>
```
Comment (uddipta) [inline Reporting.cpp]: this one too

File: SenderThread.cpp [hunk 14]
```diff
@@ -6,14 +6,14 @@
...
  */
-#include <wdt/SenderThread.h>
 #include <folly/Bits.h>
 #include <folly/Checksum.h>
 #include <folly/Conv.h>
 #include <folly/Memory.h>
 #include <folly/String.h>
 #include <sys/stat.h>
 #include <wdt/Sender.h>
+#include <wdt/SenderThread.h>
```
Comment (uddipta) [inline SenderThread.cpp]: another one

File: Throttler.cpp [hunk 15]
```diff
@@ -6,8 +6,8 @@
  * LICENSE file in the root directory of this source tree. An additional grant
  * of patent rights can be found in the PATENTS file in the same directory.
  */
-#include <wdt/Throttler.h>
 #include <wdt/ErrorCodes.h>
+#include <wdt/Throttler.h>
```
Comment (uddipta) [inline Throttler.cpp]: this one too

File: WdtOptions.cpp [hunk 16]
```diff
@@ -6,8 +6,8 @@
  * LICENSE file in the root directory of this source tree. An additional grant
  * of patent rights can be found in the PATENTS file in the same directory.
  */
-#include <wdt/WdtOptions.h>
 #include <glog/logging.h>
+#include <wdt/WdtOptions.h>
```
Comment (uddipta) [inline WdtOptions.cpp]: another

File: WdtTransferRequest.cpp [hunk 17]
```diff
@@ -6,9 +6,9 @@
  * LICENSE file in the root directory of this source tree. An additional grant
  * of patent rights can be found in the PATENTS file in the same directory.
  */
-#include <wdt/WdtTransferRequest.h>
 #include <folly/Conv.h>
 #include <folly/Range.h>
+#include <wdt/WdtTransferRequest.h>
```
Comment (uddipta) [inline WdtTransferRequest.cpp]: this one too

File: test/wdt_e2e_simple_test.sh [hunk 18]
```diff
@@ -49,6 +49,15 @@ if [ -z "$WDT_TEST_SYMLINKS" ] ; then
 fi
 echo "WDT_TEST_SYMLINKS=$WDT_TEST_SYMLINKS"
 
+# Set WDT_TEST_PERMISSION:
+# to 1 : check if permissions match
+# to 0 : when "find -printf %m" does not work, i.e. on Mac
+if [ -z "$WDT_TEST_PERMISSION" ] ; then
+  WDT_TEST_PERMISSION=1
+  umask 0
```
Comment (uddipta) [inline test/wdt_e2e_simple_test.sh]: why do we need to set this?

Comment (dliang36) [inline test/wdt_e2e_simple_test.sh]: I guess you are asking about why umask? Some machines (including mine) have a mask by default, which prevents giving write permission to Others. If we do not umask it, the file created on the receiver side will not have the write perm even if it is a 0777 on the writer side.

File: ReceiverThread.cpp (hunk 12 above)
Comment (dliang36) [inline ReceiverThread.cpp]: Ah... this is what clang gave me... Let me revert it...

**Pull request closed without merging**